server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistSnapshotsStep.java

```diff
@@ -61,104 +68,135 @@ public PersistSnapshotsStep(System2 system2, DbClient dbClient, TreeRootHolder t
+      commonForAnyVisit(directory, path, snapshot);
+    }
 
-      if (parentSnapshot != null) {
-        snapshotDto
-          .setParentId(parentSnapshot.getId())
-          .setRootId(parentSnapshot.getRootId() == null ? parentSnapshot.getId() : parentSnapshot.getRootId())
-          .setDepth(parentSnapshot.getDepth() + 1)
-          .setPath(parentSnapshot.getPath() + parentSnapshot.getId() + ".");
-      } else {
-        snapshotDto
-          // On Oracle, the path will be null
-          .setPath("")
-          .setDepth(0);
-      }
-      dbClient.snapshotDao().insert(dbSession, snapshotDto);
-      return snapshotDto;
+    @Override
+    public void visitFile(Component file, Path<SnapshotDtoHolder> path) {
+      SnapshotDto snapshot = createSnapshot(file, path, getFileQualifier(file), Scopes.FILE, false, false);
+      commonForAnyVisit(file, path, snapshot);
     }
 
-    private void updateSnapshotPeriods(SnapshotDto snapshotDto) {
-      for (Period period : periodsHolder.getPeriods()) {
-        int index = period.getIndex();
-        snapshotDto.setPeriodMode(index, period.getMode());
-        snapshotDto.setPeriodParam(index, period.getModeParameter());
-        snapshotDto.setPeriodDate(index, period.getSnapshotDate());
+    @Override
+    public void visitView(Component view, Path<SnapshotDtoHolder> path) {
+      this.rootId = dbIdsRepository.getComponentId(view);
+      SnapshotDto snapshot = createSnapshot(view, path, Qualifiers.VIEW, Scopes.PROJECT, false, true);
+      commonForAnyVisit(view, path, snapshot);
+    }
+
+    @Override
+    public void visitSubView(Component subView, Path<SnapshotDtoHolder> path) {
+      SnapshotDto snapshot = createSnapshot(subView, path, Qualifiers.SUBVIEW, Scopes.PROJECT, false, true);
+      commonForAnyVisit(subView, path, snapshot);
+    }
+
+    @Override
+    public void visitProjectView(Component projectView, Path<SnapshotDtoHolder> path) {
+      SnapshotDto snapshot = createSnapshot(projectView, path, Qualifiers.PROJECT, Scopes.FILE, false, true);
+      commonForAnyVisit(projectView, path, snapshot);
+    }
+
+    private void commonForAnyVisit(Component project, Path<SnapshotDtoHolder> path, SnapshotDto snapshot) {
+      persist(snapshot, dbSession);
+      addToCache(project, snapshot);
+      if (path.current() != null) {
+        path.current().setSnapshotDto(snapshot);
       }
     }
 
-    private void addToCache(Component component, SnapshotDto snapshotDto) {
-      dbIdsRepository.setSnapshotId(component, snapshotDto.getId());
+    private SnapshotDto createSnapshot(Component component, PathAwareVisitor.Path<SnapshotDtoHolder> path,
+      String qualifier, String scope, boolean setVersion, boolean setPeriods) {
+      return PersistSnapshotsStep.this.createSnapshot(component, path, rootId, analysisDate, qualifier, scope, setVersion, setPeriods);
+    }
+  }
+
+  private SnapshotDto createSnapshot(Component component, PathAwareVisitor.Path<SnapshotDtoHolder> path,
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Method has 8 parameters, which is greater than 7 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00107)